x-pack/libbeat/management/blacklist.go

```diff
@@ -0,0 +1,188 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package management
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/libbeat/management/api"
+)
+
+// ConfigBlacklist takes a ConfigBlocks object and filter it based on the given
+// blacklist settings
+type ConfigBlacklist struct {
+	patterns map[string]*regexp.Regexp
+}
+
+// ConfigBlacklistSettings holds a list of fields and regular expressions to blacklist
+type ConfigBlacklistSettings map[string]string
+
+// Unpack unpacks nested fields set with dot notation like foo.bar into the proper nesting
+// in a nested map/slice structure.
+func (f ConfigBlacklistSettings) Unpack(to interface{}) error {
+	m, ok := to.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("wrong type, expect map")
+	}
+
+	var expand func(key string, value interface{})
+
+	expand = func(key string, value interface{}) {
+		switch v := value.(type) {
+		case map[string]interface{}:
+			for k, val := range v {
+				expand(fmt.Sprintf("%v.%v", key, k), val)
+			}
+		case []interface{}:
+			for i := range v {
+				expand(fmt.Sprintf("%v.%v", key, i), v[i])
+			}
+		default:
+			m[key] = fmt.Sprintf("%s", value)
+		}
+	}
+
+	for k, val := range m {
+		expand(k, val)
+	}
+	return nil
+}
+
+// NewConfigBlacklist filters configs from CM according to a given blacklist
+func NewConfigBlacklist(patterns ConfigBlacklistSettings) (*ConfigBlacklist, error) {
+	list := ConfigBlacklist{
+		patterns: map[string]*regexp.Regexp{},
+	}
+
+	for field, pattern := range patterns {
+		exp, err := regexp.Compile(pattern)
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("Given expression is not a valid regexp: %s", pattern))
+		}
+
+		list.patterns[field] = exp
+	}
+
+	return &list, nil
+}
+
+// Filter returns a copy of the given ConfigBlocks with the
+func (c *ConfigBlacklist) Filter(configBlocks api.ConfigBlocks) api.ConfigBlocks {
+	var result api.ConfigBlocks
+
+	for _, configs := range configBlocks {
+		newConfigs := api.ConfigBlocksWithType{Type: configs.Type}
+
+		for _, block := range configs.Blocks {
+			if c.blacklisted(configs.Type, block) {
+				logp.Err("Got a blacklisted configuration, ignoring it")
+				continue
+			}
+
+			newConfigs.Blocks = append(newConfigs.Blocks, block)
+		}
+
+		if len(newConfigs.Blocks) > 0 {
+			result = append(result, newConfigs)
+		}
+	}
+
+	return result
+}
```

Comment:
From what I see, lets say that we receive a config block with a black listed option, we would return log an error and still apply the other config blocks? 

I think to have least surprise as a user it would be better to refuse to apply _all the blocks_, my reasoning is if we instead apply blocks in part we have an inconsistent view of the configuration and this would be unexpected? 

Maybe I am wrong.

Reply:
Or maybe we don't have the mechanism yet in place to communicate that problem to the user through the UI?

Reply:
This is a fair question, I don't know what the user should expect in these cases. So your approach would be to "pause" the beat, and disable all modules & outputs until the configuration is not blacklisted?

We currently have a way to report an error, it refers users to the log file when that happens.

Reply:
No matter what, we also to signal the UI that configs can not be applied due to the blacklist. Just logging the issue locally, makes it somewhat 'silent'.

Reply:
Also, people could have access to management but don't have physical access to the machine. 

@urso I think it could be done in two PR, this one to detect errors and another one to improve the feedback loop of a beats since this would certainly require UI changes to support it.